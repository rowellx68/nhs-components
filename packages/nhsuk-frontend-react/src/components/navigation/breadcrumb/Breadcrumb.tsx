import React from 'react';
import { VisuallyHidden } from '@/components/core/visually-hidden/VisuallyHidden';
import { Base, BaseProps } from '@/components/core/base/Base';
import {
  polymorphicFactory,
  PolymorphicFactory,
} from '@/internal/factory/polymorphic-factory';
import clsx from 'clsx';
import {
  AsElementProps,
  ElementProps,
  WithVisuallyHiddenTextProps,
} from '@/types/shared';
import { Factory, factory } from '@/internal/factory/factory';

export type BreadcrumbProps = ElementProps<'nav'>;

type BreadcrumbFactory = Factory<{
  props: BreadcrumbProps;
  ref: HTMLDivElement;
  staticComponents: {
    List: typeof BreadcrumbList;
    ListItem: typeof BreadcrumbListItem;
    BackLink: typeof BreadcrumbBackLink;
  };
}>;

const Breadcrumb = factory<BreadcrumbFactory>(
  ({ children, className, ...props }: BreadcrumbProps, ref) => {
    return (
      <nav
        className={clsx('nhsuk-breadcrumb', className)}
        aria-label="Breadcrumb"
        {...props}
        ref={ref}
      >
        {children}
      </nav>
    );
  },
);

export type BreadcrumbListProps = ElementProps<'ol', 'type'>;

const BreadcrumbList = ({
  children,
  className,
  ...props
}: BreadcrumbListProps) => {
  return (
    <ol className={clsx('nhsuk-breadcrumb__list', className)} {...props}>
      {children}
    </ol>
  );
};

export type BreadcrumbListItemProps = BaseProps;

type BaseCrumpListItemFactory = PolymorphicFactory<{
  props: BreadcrumbListItemProps;
  defaultComponent: 'a';
  defaultRef: HTMLAnchorElement;
}>;

const BreadcrumbListItem = polymorphicFactory<BaseCrumpListItemFactory>(
  (
    {
      className,
      as: component = 'a',
      ...props
    }: BreadcrumbListItemProps & AsElementProps,
    ref,
  ) => {
    return (
      <li className="nhsuk-breadcrumb__item">
        <Base
          as={component}
          className={clsx('nhsuk-breadcrumb__link', className)}
          {...props}
          ref={ref}
        />
      </li>
    );
  },
);
export type BreadcrumbBackLinkProps = BaseProps & WithVisuallyHiddenTextProps;

type BreadcrumbBackLinkFactory = PolymorphicFactory<{
  props: BreadcrumbBackLinkProps;
  defaultComponent: 'a';
  defaultRef: HTMLAnchorElement;
}>;

const BreadcrumbBackLink = polymorphicFactory<BreadcrumbBackLinkFactory>(
  (
    {
      className,
      children,
      visuallyHiddenText = 'Back to &nbsp;',
      as: component = 'a',
      ...props
    }: BreadcrumbBackLinkProps & AsElementProps,
    ref,
  ) => {
    return (
      <p className="nhsuk-breadcrumb__back">
        <Base
          as={component}
          className={clsx('nhsuk-breadcrumb__backlink', className)}
          {...props}
          ref={ref}
        >
          <VisuallyHidden>{visuallyHiddenText}</VisuallyHidden>
          {children}
        </Base>
      </p>
    );
  },
);

Breadcrumb.displayName = 'Breadcrumb';
BreadcrumbList.displayName = 'Breadcrumb.List';
BreadcrumbListItem.displayName = 'Breadcrumb.ListItem';
BreadcrumbBackLink.displayName = 'Breadcrumb.BackLink';

Breadcrumb.List = BreadcrumbList;
Breadcrumb.ListItem = BreadcrumbListItem;
Breadcrumb.BackLink = BreadcrumbBackLink;

export { Breadcrumb, BreadcrumbList, BreadcrumbListItem, BreadcrumbBackLink };
