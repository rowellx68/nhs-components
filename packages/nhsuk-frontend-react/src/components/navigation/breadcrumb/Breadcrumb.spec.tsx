import React from 'react';
import { it, expect } from 'vitest';
import { render } from '@testing-library/react';
import { Breadcrumb } from './Breadcrumb';

it('should render a breadcrumb element', () => {
  const { container } = render(
    <Breadcrumb>
      <Breadcrumb.List>
        <Breadcrumb.ListItem href="/level-one">Level one</Breadcrumb.ListItem>
        <Breadcrumb.ListItem href="/level-one/level-two">
          Level two
        </Breadcrumb.ListItem>
        <Breadcrumb.ListItem href="/level-one/level-two/level-three">
          Level three
        </Breadcrumb.ListItem>
      </Breadcrumb.List>
      <Breadcrumb.BackLink href="/level-one/level-two/level-three">
        Level three
      </Breadcrumb.BackLink>
    </Breadcrumb>,
  );

  expect(container.querySelector('.nhsuk-breadcrumb')).toBeInTheDocument();
  expect(container.querySelectorAll('a')).toHaveLength(4);
  expect(container).toMatchSnapshot();
});
