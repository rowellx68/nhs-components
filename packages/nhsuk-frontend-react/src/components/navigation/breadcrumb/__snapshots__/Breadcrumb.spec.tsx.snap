// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render a breadcrumb element 1`] = `
<div>
  <nav
    aria-label="Breadcrumb"
    class="nhsuk-breadcrumb"
  >
    <div
      class="nhsuk-width-container"
    >
      <ol
        class="nhsuk-breadcrumb__list"
      >
        <li
          class="nhsuk-breadcrumb__item"
        >
          <a
            class="nhsuk-breadcrumb__link"
            href="/level-one"
          >
            Level one
          </a>
        </li>
        <li
          class="nhsuk-breadcrumb__item"
        >
          <a
            class="nhsuk-breadcrumb__link"
            href="/level-one/level-two"
          >
            Level two
          </a>
        </li>
        <li
          class="nhsuk-breadcrumb__item"
        >
          <a
            class="nhsuk-breadcrumb__link"
            href="/level-one/level-two/level-three"
          >
            Level three
          </a>
        </li>
      </ol>
      <p
        class="nhsuk-breadcrumb__back"
      >
        <a
          class="nhsuk-breadcrumb__backlink"
          href="/level-one/level-two/level-three"
        >
          <span
            class="nhsuk-u-visually-hidden"
          >
            Back to &nbsp;
          </span>
          Level three
        </a>
      </p>
    </div>
  </nav>
</div>
`;
